test/ports_test.go

```diff
@@ -0,0 +1,185 @@
+
+	s := RunServer(&opts)
+	if !s.ReadyForListeners(5 * time.Second) {
+		t.Fatal("services failed to start in 5 seconds")
+	}
+
+	// the pid file should be
+	portsFile := s.PortsFile(opts.PortsFileDir)
+
+	if portsFile == "" {
+		t.Fatal("Expected a ports file")
+	}
+
+	// try to read a file here
+
+	// the file should be a json
+	buf, err := waitForFile(portsFile, 5*time.Second)
+	if err != nil {
+		t.Fatalf("Could not read ports file: %v", err)
+	}
+	if len(buf) <= 0 {
+		t.Fatal("Expected a non-zero length ports file")
+	}
+
+	ports := server.Ports{}
+	json.Unmarshal(buf, &ports)
+
+	if len(ports.Nats) == 0 || !strings.HasPrefix(ports.Nats[0], "nats://") {
+		t.Fatal("Expected at least one nats url")
+	}
+
+	if len(ports.Monitoring) == 0 || !strings.HasPrefix(ports.Monitoring[0], "http://") {
+		t.Fatal("Expected at least one monitoring url")
+	}
+
+	if len(ports.Cluster) == 0 || !strings.HasPrefix(ports.Cluster[0], "nats://") {
+		t.Fatal("Expected at least one cluster listen url")
+	}
+
+	if len(ports.Profile) == 0 || !strings.HasPrefix(ports.Profile[0], "http://") {
+		t.Fatal("Expected at least one profile listen url")
+	}
+
+	s.Shutdown()
+
+	// if we called shutdown, the cleanup code should have kicked
+	if _, err := os.Stat(portsFile); os.IsNotExist(err) {
+		// good
+	} else {
+		t.Fatalf("the port file %s was not deleted", portsFile)
+	}
+}
+
+// makes a temp directory with two directories 'A' and 'B'
+// the location of the ports file is changed from dir A to dir B.
+func TestPortsFileReload(t *testing.T) {
+	// make a temp dir
+	tempDir, err := ioutil.TempDir("", "")
+	if err != nil {
+		t.Fatalf("Error creating temp director (%s): %v", tempDir, err)
+	}
+
+	// make child temp dir A
+	dirA := path.Join(tempDir, "A")
+	os.MkdirAll(dirA, 0777)
+
+	// write the config file with a reference to A
+	config := fmt.Sprintf("ports_file_dir %s\nport 1966", dirA)
+	confPath := path.Join(tempDir, fmt.Sprintf("%d.conf", os.Getpid()))
+	if err := ioutil.WriteFile(confPath, []byte(config), 0666); err != nil {
+		t.Fatalf("Error writing ports file (%s): %v", confPath, err)
+	}
+
+	opts, err := server.ProcessConfigFile(confPath)
+	if err != nil {
+		t.Fatalf("Error processing the configuration: %v", err)
+	}
+
+	s := RunServer(opts)
+	if !s.ReadyForListeners(5 * time.Second) {
+		t.Fatal("services failed to start in 5 seconds")
+	}
+	defer func() {
+		s.Shutdown()
+	}()
+
+	// get the ports file path name
+	portsFileInA := s.PortsFile("")
+	// the file should be in dirA
+	if !strings.HasPrefix(portsFileInA, dirA) {
+		t.Fatalf("expected ports file to be in [%s] but was in [%s]", dirA, portsFileInA)
+	}
+	// wait for it
+	buf, err := waitForFile(portsFileInA, 5*time.Second)
+	if err != nil {
+		t.Fatalf("Could not read ports file: %v", err)
+	}
+	if len(buf) <= 0 {
+		t.Fatal("Expected a non-zero length ports file")
+	}
+
+	// change the configuration for the ports file to dirB
+	dirB := path.Join(tempDir, "B")
+	os.MkdirAll(dirB, 0777)
+
```

Comment:
Add `defer os.RemoveAll(dirB)`

Reply:
we just need to delete the tempDir - both A and B are in it.